src/tensorwaves/estimator.py

```diff
@@ -2,24 +2,63 @@
 
 All estimators have to implement the `~.interfaces.Estimator` interface.
 """
+from typing import Dict
+
 import numpy as np
 import tensorflow as tf
 
 from tensorwaves.interfaces import Estimator, Function
 
 
+class _NormalizedFunction(Function):
+    def __init__(
+        self,
+        unnormalized_function: Function,
+        norm_dataset: dict,
+        norm_volume: float = 1.0,
+    ) -> None:
+        self._model = unnormalized_function
+        # it is crucial to convert the input data to tensors
+        # otherwise the memory footprint can increase dramatically
+        self._norm_dataset = {
+            x: tf.constant(y) for x, y in norm_dataset.items()
+        }
+        self._norm_volume = norm_volume
+
+    def __call__(self, dataset: dict) -> tf.Tensor:
+        normalization = tf.multiply(
+            self._norm_volume,
+            tf.reduce_mean(self._model(self._norm_dataset)),
+        )
+        return tf.divide(self._model(dataset), normalization)
+
+    @property
+    def parameters(self) -> Dict[str, tf.Variable]:
+        return self._model.parameters
+
+    def update_parameters(self, new_parameters: dict) -> None:
+        self._model.update_parameters(new_parameters)
+
+
 class UnbinnedNLL(Estimator):
     """Unbinned negative log likelihood estimator.
 
     Args:
         model: A model that should be compared to the dataset.
         dataset: The dataset used for the comparison. The model has to be
             evaluateable with this dataset.
+        phsp_set: A phase space dataset, which is used for the normalization.
+            The model has to be evaluateable with this dataset. When correcting
+            for the detector efficiency use a phase space sample, that passed
+            the detector reconstruction.
 
     """
 
-    def __init__(self, model: Function, dataset: dict) -> None:
-        self.__model = model
+    def __init__(self, model: Function, dataset: dict, phsp_set: dict) -> None:
```

Comment:
If I understand correctly, this follows the interface of ComPWA, see here
https://compwa.github.io/ComPWA/classComPWA_1_1Estimator_1_1MinLogLH.html

What was the reason not to have `phsp_set` as a required argument for this estimator in `tensorwaves`?

Reply:
Good quesion, unfortunately I did not write down my reasoning back then. I guess simplicity, for the estimator. Naively speaking an estimator is supposed to compare a function to a dataset. This phase space sample is not exactly intuative.